### DISCUSSION TITLE
Speech Recognition on Android

### DISCUSSION THREAD

**Discussion by wagenheimer** [shinyorg/shiny#708] [Q&A]
Anybody here was able to make Speech Recognition work on Android as expected?

I implemented everything as expected, but both Listen for a Phrase and Continuous Dictation do have problems. 

- I need to capture single answers (yes, no, or numbers), but using "Listen for a phrase" does not hear what I say 90% of the time.
- Continues Dictation works randomly, it does not hear most of what I say, but something it hears full phrases correctly.

Before using Shiny, I tried to do it manually using this code, and It worked:

```
public void StartSpeechToText(ISpeechResultCallback callback)
    {

        string rec = global::Android.Content.PM.PackageManager.FeatureMicrophone;
        if (rec == "android.hardware.microphone")
        {
            var voiceIntent = new Intent(RecognizerIntent.ActionRecognizeSpeech);
            voiceIntent.PutExtra(RecognizerIntent.ExtraLanguageModel, RecognizerIntent.LanguageModelFreeForm);

            voiceIntent.PutExtra(RecognizerIntent.ExtraPrompt, "Speak now");

            voiceIntent.PutExtra(RecognizerIntent.ExtraSpeechInputCompleteSilenceLengthMillis, 1500);
            voiceIntent.PutExtra(RecognizerIntent.ExtraSpeechInputPossiblyCompleteSilenceLengthMillis, 1500);
            voiceIntent.PutExtra(RecognizerIntent.ExtraSpeechInputMinimumLengthMillis, 15000);
            voiceIntent.PutExtra(RecognizerIntent.ExtraMaxResults, 1);
            voiceIntent.PutExtra(RecognizerIntent.ExtraLanguage, Java.Util.Locale.Default);

            activity.StartActivityForResult(voiceIntent, VOICE);
        }
    }

   protected override void OnActivityResult(int requestCode, Result resultCode, Intent data)
        {

            if (requestCode == VOICE)
            {
                if (resultCode == Result.Ok)
                {
                    var matches = data.GetStringArrayListExtra(RecognizerIntent.ExtraResults);
                    if (matches.Count != 0)
                    {
                        var textInput = matches[0];
                        MessagingCenter.Send<IMessageSender, string>(this, "STT", textInput);
                    }
                    else
                    {
                        MessagingCenter.Send<IMessageSender, string>(this, "STT", "No input");
                    }

                }
            }
            base.OnActivityResult(requestCode, resultCode, data);
        }
    }
```

It understands what I say almost 100% of the time. But it does show a dialog when it's listening, and I want to avoid that dialog so I can automatically turn on listening without user interaction.

Shiny does this, but unfortunately for some reason, when I use it, it doesn't understand or hear what I say. It makes the listening sound, I speak, then it makes the "error" sound. Sometimes when it understands, it makes the sound of "success" and correctly understands what I say.

I checked all the Shiny source code on Git, and in the part that I understood, I couldn't see anything different from the code that I manually tested and it worked.

Is anyone successfully using it with Android?

**Comment by vovanb** [comment]
your issue with ExtraSpeechInputCompleteSilenceLengthMillis - once you use it, recognition does not work
